### ISSUE TITLE
[BUG] Twilio Device gets into an inconsistent state after trouble re-connecting to the signaling websocket

### ISSUE THREAD

**Issue by acnovais** [twilio/twilio-client.js#227]
- [x] I have verified that the issue occurs with the latest twilio.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-client.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-client.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/TwilioDevEd/client-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

This error was identified when I saw evidence of some users of my dialpad not receiving any more calls after experiencing connectivity issues and re-connecting. Furthermore, the calls status from twilio reported as having ringed and timedout (no-answer). My investigation lead to the conclusion that:
* when the user is with an active call
* and loses the connection to the signaling websocket
* cannot reconnect for a few retries
* Can finally reconnect after

The twilio sdk gets to an inconsistent state and thinks that it still holds an active connection. The effect is that new incoming calls are ignored (we set allowIncomingWhileBusy to false) and the client application is unaware of these calls.
The user needs to reload the page in order to fix his issue.

**How to reproduce the issue:**
Pre-requirement
```
//Setup twilio with allowIncomingWhileBusy set to false
twilioDevice.setup(token, { allowIncomingWhileBusy: false, debug: true })
```
Run your code in development

1. set a breakpoint on this [line](https://github.com/twilio/twilio-client.js/blob/2a7f457b7efe15684622bdd292db07d4228ff2f4/lib/twilio/wstransport.ts#L296)
2. force the connection to twilio (debugger will pause there)
3. step over so that the websocket is instantiated and make sure you save a reference to it globally (e.g. window.temp1) and resume execution
4. establish a call (inbound or outbound) and accept it
5. edit your `/etc/hosts` file so that the client sdk won't be able to reconnect to the websocket server when it tries to reconnect (e.g. `127.0.0.1 chunderw-vpc-gll.twilio.com`)
6. Force the closing of the websocket (e.g. `window.temp1.close()`)
7. Wait for a few re-connection retries to fail
8. remove the rule on the `/etc/hosts` file
9. wait for the twilio device to have successfully reconnected
10. place an incoming call to this device
11. You'll see a log message with [Device busy; ignoring incoming invite](https://github.com/twilio/twilio-client.js/blob/0eced788a786880f83c51ddee3ce48b6fbb5b365/lib/twilio/device.ts#L1113)

**Expected behavior:**

The internal sdk state should have been correctly reset and the new calls should be delivered to the client app

**Actual behavior:**

The incoming calls are ignored and not delivered to the client app

**Software versions:**

 - [x] Browser(s): Chrome 85
 - [x] Operating System: macOS Catalina
 - [x] twilio.js: 1.12.5
 - [ ] Third-party libraries (e.g., Angular, React, etc.):

**Comment by charliesantos** [comment]
Thanks for submitting. We'll take a look.
JIRA: https://issues.corp.twilio.com/browse/CLIENT-8116

**Comment by charliesantos** [comment]
Hi @acnovais ,
The reason why this is happening is because ICE will try to re-negotiate after a successful reconnect (step 9) and will try to recover the media. In your scenario, the client is still busy recovering ICE, which explains why incoming calls are ignored. But ICE should timeout after a few seconds which will disconnect the connection automatically and raises a 31003 error. After the connection is disconnected, the client should be able to receive incoming calls again.

If you have Device options `enableIceRestart` set to true, the reconnection times will take longer in this specific use case. Do you have it enabled by any chance?

Thank you,
Charlie

**Comment by acnovais** [comment]
Hi @charliesantos, thanks for looking into it.

Looking a bit further into this, what I realize is that when this happens Twilio's perception of the Call having ended is different from what the webRTC internals is reporting. Twilio actually reported that the call ended between steps 6 and 8, but the actual webRTC connection was still up, and only when I forced the disconnection on twilio SDK did it close (screenshots bellow where you can see the different closed/ended connection timestamps). This causes our server to be out of sync with what is happening with the client, and therefore new calls are routed to this client when it is in fact busy (the webrtc connection is still live)

![image](https://user-images.githubusercontent.com/930394/97040358-b9eb0700-1565-11eb-8d76-aeee3971e715.png)


![image](https://user-images.githubusercontent.com/930394/97040429-d0915e00-1565-11eb-8bbf-09c7522825d2.png)

**Comment by charliesantos** [comment]
Hi @acnovais , that is correct. WebRTC connection, the ICE connection specifically, still thinks it's connected and will try to recover. But this will time out eventually, which will disconnect the connection. If you try to wait several seconds after step 9, you should see a 31003 error. After that, you should be able to receive incoming calls. Can you please try this? That is, wait for the timeout.
Also, is Device options enableIceRestart set to true? This setting will affect this reconnection behavior, but I don't want to go that route if you don't have it set. Can you please let me know if you are setting this?

**Comment by acnovais** [comment]
I actually waited over 1minute and the webrtc connection remained active. Only when I manually did `.disconnect()` is when the connection truly closed (that is visible on the screenshot). No, I'm not using any other option besides the ones mentioned on the issue.

Thanks again for your help 🙏

**Comment by charliesantos** [comment]
Hi @acnovais ,
Thanks for the additional information. Ideally in this situation, the SDK should reconnect the call automatically after the websocket is disconnected. However, that is currently in development and might take time due to the required changes on the server.

As a workaround, you can listen for `connection.on('transportClose', handler)` then disconnect the connection. That way, you don't get into this weird state. For example:
```
connection.on('transportClose', () => {
    connection.disconnect();
});
```
Please keep in mind that this `transportClose` event is raised whenever websocket is disconnected which usually happens when there's network issues, or when you switch to a different network. If you are ok with this behavior, please try it.
